# Pandas CSV 文件

CSV(Comma-Separated Values, 逗号分隔值，有时称字符分隔值，即分隔符可以不是逗号)，其文件以纯文本形式纯村表格数据（数字和文本）。


### 1. pd.read_csv() -- 读取CSV文件

```pytohn
# 返回DataFrame
df = pd.read_csv('data.csv', sep=';', header=0, names=['A', 'B', 'C'], dtype={'A': int, 'B': float})
```

| 参数                | 说明                                                         | 默认值   |
|---------------------|--------------------------------------------------------------|----------|
| filepath_or_buffer  | CSV 文件的路径或文件对象（支持 URL、文件路径、文件对象等）| 必需参数 |
| sep                 | 定义字段分隔符，默认是逗号（,），可以改为其他字符，如制表符（\t） | ','      |
| header              | 指定行号作为列标题，默认为 0（表示第一行），或者设置为 None 没有标题 | 0        |
| names               | 自定义列名，传入列名列表                                     | None     |
| index_col           | 用作行索引的列的列号或列名                                   | None     |
| usecols             | 读取指定的列，可以是列的名称或列的索引                       | None     |
| dtype               | 强制将列转换为指定的数据类型                                 | None     |
| skiprows            | 跳过文件开头的指定行数，或者传入一个行号的列表               | None     |
| nrows               | 读取前 N 行数据                                              | None     |
| na_values           | 指定哪些值应视为缺失值（NaN）| None     |
| skipfooter          | 跳过文件结尾的指定行数                                       | 0        |
| encoding            | 文件的编码格式（如 utf-8，latin1 等）| None     |

`display(df.to_string())`: 返回DataFrame, 如果不使用，则输出结果为数据的前5行和末尾5行，中间以`...`代替

In [3]:
import pandas as pd

df = pd.read_csv('nba.csv')
display(df)
display(df.to_string())

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN
3,R.J. Hunter,Boston Celtics,28.0,SG,22.0,6-5,185.0,Georgia State,1148640.0
4,Jonas Jerebko,Boston Celtics,8.0,PF,29.0,6-10,231.0,NaN,5000000.0
...,...,...,...,...,...,...,...,...,...
453,Shelvin Mack,Utah Jazz,8.0,PG,26.0,6-3,203.0,Butler,2433333.0
454,Raul Neto,Utah Jazz,25.0,PG,24.0,6-1,179.0,NaN,900000.0
455,Tibor Pleiss,Utah Jazz,21.0,C,26.0,7-3,256.0,NaN,2900000.0
456,Jeff Withey,Utah Jazz,24.0,C,26.0,7-0,231.0,Kansas,947276.0


"                         Name                    Team  Number Position   Age Height  Weight                College      Salary\n0               Avery Bradley          Boston Celtics     0.0       PG  25.0    6-2   180.0                  Texas   7730337.0\n1                 Jae Crowder          Boston Celtics    99.0       SF  25.0    6-6   235.0              Marquette   6796117.0\n2                John Holland          Boston Celtics    30.0       SG  27.0    6-5   205.0      Boston University         NaN\n3                 R.J. Hunter          Boston Celtics    28.0       SG  22.0    6-5   185.0          Georgia State   1148640.0\n4               Jonas Jerebko          Boston Celtics     8.0       PF  29.0   6-10   231.0                    NaN   5000000.0\n5                Amir Johnson          Boston Celtics    90.0       PF  29.0    6-9   240.0                    NaN  12000000.0\n6               Jordan Mickey          Boston Celtics    55.0       PF  21.0    6-8   235.0            

### 2. pd.to_csv() -- 将DataFrame写入CSV文件

`df.to_csv('output.csv', index=False, header=True, columns=['A', 'B'])`

| 参数              | 说明                                                         | 默认值   |
|-------------------|--------------------------------------------------------------|----------|
| path_or_buffer    | CSV 文件的路径或文件对象（支持文件路径、文件对象）| 必需参数 |
| sep               | 定义字段分隔符，默认是逗号（,），可以改为其他字符，如制表符（\t） | ','      |
| index             | 是否写入行索引，默认 True 表示写入索引                       | True     |
| columns           | 指定写入的列，可以是列的名称列表                             | None     |
| header            | 是否写入列名，默认 True 表示写入列名，设置为 False 表示不写列名 | True     |
| mode              | 写入文件的模式，默认是 w（写模式），可以设置为 a（追加模式） | 'w'      |
| encoding          | 文件的编码格式，如 utf-8，latin1 等                          | None     |
| line_terminator   | 定义行结束符，默认为 \n                                      | None     |
| quoting           | 设置如何对文件中的数据进行引号处理（0 - 3，具体引用方式可查文档） | None     |
| quotechar         | 设置用于引用的字符，默认为双引号 "                            | '"'      |
| date_format       | 自定义日期格式，如果列包含日期数据，则可以使用此参数指定日期格式 | None     |
| doublequote       | 如果为 True，则在写入时会将包含引号的文本使用双引号括起来   | True     |

In [4]:
import pandas as pd 
   
# 三个字段 name, site, age
nme = ["Google", "Runoob", "Taobao", "Wiki"]
st = ["www.google.com", "www.runoob.com", "www.taobao.com", "www.wikipedia.org"]
ag = [90, 40, 80, 98]
   
# 字典
dict = {'name': nme, 'site': st, 'age': ag} 
     
df = pd.DataFrame(dict)
  
# 保存 dataframe
df.to_csv('site.csv')

In [5]:
import pandas as pd

df = pd.read_csv('nba.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


在 Pandas 中，`object` 是 DataFrame 中一种灵活的数据类型，通常用于存储以下内容：

1. **字符串数据**：当列中包含文本信息（如名称、描述等）时，Pandas 会默认将其识别为 `object` 类型。
   
2. **混合类型数据**：如果一列中同时包含多种数据类型（如数字、字符串、None 等），Pandas 无法将其统一为更具体的类型（如 int、float），会将其标记为 `object` 类型。

3. **Python 对象**：偶尔也会用于存储 Python 中的其他对象（如自定义类的实例），但这种情况较少见。

例如，一个包含姓名、地址的列通常是 `object` 类型；如果一列中既有整数 `123` 又有字符串 `"456"`，也会被识别为 `object` 类型。

`object` 类型的灵活性较高，但运算效率通常低于数值类型（如 int、float），因此在数据分析中，有时需要将其转换为更合适的类型（如 `astype(str)` 显式转为字符串，或通过 `pd.to_numeric()` 转为数值类型）。

# Pandas Excel 文件操作

| 操作                     | 方法                | 说明                                       |
|--------------------------|---------------------|--------------------------------------------|
| 读取 Excel 文件          | pd.read_excel()     | 读取 Excel 文件，返回 DataFrame             |
| 将 DataFrame 写入 Excel  | DataFrame.to_excel()| 将 DataFrame 写入 Excel 文件                |
| 加载 Excel 文件          | pd.ExcelFile()      | 加载 Excel 文件并访问多个表单               |
| 使用 ExcelWriter 写多个表单 | pd.ExcelWriter()    | 写入多个 DataFrame 到同一 Excel 文件的不同表单 |

### 1. pd.read_excel() -- 读取 Excel 文件

`pd.read_excel(io, sheet_name=0, *, header=0, names=None, index_col=None, usecols=None, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, parse_dates=False, date_parser=<no_default>, date_format=None, thousands=None, decimal='.', comment=None, skipfooter=0, storage_options=None, dtype_backend=<no_default>, engine_kwargs=None)`

参数说明：

`io`：这是必需的参数，指定了要读取的 Excel 文件的路径或文件对象。

`sheet_name=0`：指定要读取的工作表名称或索引。默认为0，即第一个工作表。

`header=0`：指定用作列名的行。默认为0，即第一行。

`names=None`：用于指定列名的列表。如果提供，将覆盖文件中的列名。

`index_col=None`：指定用作行索引的列。可以是列的名称或数字。

`usecols=None`：指定要读取的列。可以是列名的列表或列索引的列表。

`dtype=None`：指定列的数据类型。可以是字典格式，键为列名，值为数据类型。

`engine=None`：指定解析引擎。默认为None，pandas 会自动选择。

`converters=None`：用于转换数据的函数字典。

`true_values=None`：指定应该被视为布尔值True的值。

`false_values=None`：指定应该被视为布尔值False的值。

`skiprows=None`：指定要跳过的行数或要跳过的行的列表。

`nrows=None`：指定要读取的行数。

`na_values=None`：指定应该被视为缺失值的值。

`keep_default_na=True`：指定是否要将默认的缺失值（例如NaN）解析为NA。

`na_filter=True`：指定是否要将数据转换为NA。

`verbose=False`：指定是否要输出详细的进度信息。

`parse_dates=False`：指定是否要解析日期。

`date_parser=<no_default>`：用于解析日期的函数。

`date_format=None`：指定日期的格式。

`thousands=None`：指定千位分隔符。

`decimal='.'`：指定小数点字符。

`comment=None`：指定注释字符。

`skipfooter=0`：指定要跳过的文件末尾的行数。

`storage_options=None`：用于云存储的参数字典。

`dtype_backend=<no_default>`：指定数据类型后端。

`engine_kwargs=None`：传递给引擎的额外参数字典。

```python
# 读取默认的第一个表单
df = pd.read_excel('data.xlsx')
print(df)

# 读取指定表单的内容（表单名称）
df = pd.read_excel('data.xlsx', sheet_name='Sheet1')
print(df)

# 读取多个表单，返回一个字典
dfs = pd.read_excel('data.xlsx', sheet_name=['Sheet1', 'Sheet2'])
print(dfs)

# 自定义列名并跳过前两行
df = pd.read_excel('data.xlsx', header=None, names=['A', 'B', 'C'], skiprows=2)
```

In [6]:
import pandas as pd

df = pd.read_excel('runoob_pandas_data.xlsx')

df

,表格 1,Unnamed: 1,Unnamed: 2
0,Name,Age,City
1,Alice,25,New York
2,Bob,30,Los Angeles
3,Charlie,35,Chicago


### 2. df.to_excel() -- 将DataFrame写入Excel文件

`.to_excel()`方法支持 `.xls` 和 `.xlsx` 格式

`
df.to_excel(excel_writer, *, sheet_name='Sheet1', na_rep='', float_format=None, columns=None, header=True, index=True, index_label=None, startrow=0, startcol=0, engine=None, merge_cells=True, inf_rep='inf', freeze_panes=None, storage_options=None, engine_kwargs=None)
`

参数说明：

`excel_writer`：这是必需的参数，指定了要写入的 Excel 文件路径或文件对象。

`sheet_name='Sheet1'`：指定写入的工作表名称，默认为 'Sheet1'。

`na_rep=''`：指定在 Excel 文件中表示缺失值（NaN）的字符串，默认为空字符串。

`float_format=None`：指定浮点数的格式。如果为 None，则使用 Excel 的默认格式。

`columns=None`：指定要写入的列。如果为 None，则写入所有列。

`header=True`：指定是否写入列名作为第一行。如果为 False，则不写入列名。

`index=True`：指定是否写入索引作为第一列。如果为 False，则不写入索引。

`index_label=None`：指定索引列的标签。如果为 None，则不写入索引标签。

`startrow=0`：指定开始写入的行号，默认从第0行开始。

`startcol=0`：指定开始写入的列号，默认从第0列开始。

`engine=None`：指定写入 Excel 文件时使用的引擎，默认为 None，pandas 会自动选择。

`merge_cells=True`：指定是否合并单元格。如果为 True，则合并具有相同值的单元格。

`inf_rep='inf'`：指定在 Excel 文件中表示无穷大值的字符串，默认为 'inf'。

`freeze_panes=None`：指定冻结窗格的位置。如果为 None，则不冻结窗格。

`storage_options=None`：用于云存储的参数字典。

`engine_kwargs=None`：传递给引擎的额外参数字典。

In [10]:
import pandas as pd

df = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [20, 25, 30],
    'City': ['New York', 'San Francisco', 'Los Angeles']
})

df.to_excel('output.xlsx', sheet_name='Sheet1', index=False)

# 写入多个表单
with pd.ExcelWriter('output(多个表).xlsx') as writer:
    df.to_excel(writer, sheet_name='Sheet1', index=False)
    df.to_excel(writer, sheet_name='Sheet2', index=False)

### 3. pd.ExcelFile -- 加载 Excel 文件

可以处理多个表单，并不重新打开文件的情况下访问其中的数据

`excel_file = pd.ExcelFile('data.xlsx')`

| 方法 | 功能                   |
|------|----------------------|
| sheet_name | 返回文件中所有表单的名称列表       |
| parse(sheet_name) | 解析指定表单并返回一个 DataFrame |
| close() | 关闭文件，以释放资源           |



In [18]:
import pandas as pd

# 使用ExcelFile加载Excel文件
excel_file = pd.ExcelFile('output(多个表).xlsx')

# 查看所有表单名称
display(excel_file.sheet_names)

# 读取指定表单
df = excel_file.parse('Sheet2', header=2)  # 读取第三行(不算第一行的列名)
display(df)

# 关闭文件
excel_file.close()

['Sheet1', 'Sheet2']

,Bob,25,San Francisco
0,Charlie,30,Los Angeles


### 4. pd.ExcelWriter -- 写入 Excel 文件

ExcelWriter 是 pandas 提供的一个类，用于将 DataFrame 或 Series 对象写入 Excel 文件。使用 ExcelWriter，你可以在一个 Excel 文件中写入**多个**工作表，并且可以更灵活地控制写入过程。

`pd.ExcelFile(path, engine=None, data_format=None, datetime_format=None, mode='w', storage_options=None, if_sheet_exists=None, engine_kwargs=None)`

参数说明：

`path`：这是必需的参数，指定了要写入的 Excel 文件的路径、URL 或文件对象。可以是本地文件路径、远程存储路径（如 S3）、URL 链接或已打开的文件对象。

`engine`：这是一个可选参数，用于指定写入 Excel 文件的引擎。如果为 None，则 pandas 会自动选择一个可用的引擎（默认优先选择 `openpyxl`，如果不可用则选择其他可用引擎）。常见的引擎包括 `openpyxl`（用于 .xlsx 文件）、`xlsxwriter`（提供高级格式化和图表功能）、`odf`（用于 OpenDocument 格式如 .ods）等。

`date_format`：这是一个可选参数，指定写入 Excel 文件中日期的格式字符串，例如 "YYYY-MM-DD"。

`datetime_format`：这是一个可选参数，指定写入 Excel 文件中日期时间对象的格式字符串，例如 "YYYY-MM-DD HH:MM:SS"。

`mode`：这是一个可选参数，默认为 'w'，表示写入模式。如果设置为 'a'，则表示追加模式，向现有文件中添加数据（仅支持**部分引擎**，如 `openpyxl`）。

`storage_options`：这是一个可选参数，用于指定与存储后端连接的额外选项，例如认证信息、访问权限等，适用于写入远程存储（如 S3、GCS）。

`if_sheet_exists`：这是一个可选参数，默认为 `error`，指定如果工作表已经存在时的行为。选项包括 `'error'`（抛出错误）、`'new'`（创建一个新工作表）、`'replace'`（替换现有工作表的内容）、`'overlay'`（在现有工作表上覆盖写入）。

`engine_kwargs`：这是一个可选参数，用于传递给引擎的其他关键字参数。这些参数会传递给相应引擎的函数，例如 `xlsxwriter.Workbook(file, **engine_kwargs)` 或 `openpyxl.Workbook(**engine_kwargs)` 等。

In [9]:
df1 = pd.DataFrame([["AAA", "BBB"]], columns=["Spam", "Egg"])  
df2 = pd.DataFrame([["ABC", "XYZ"]], columns=["Foo", "Bar"])  
with pd.ExcelWriter("path_to_file.xlsx") as writer:
    df1.to_excel(writer, sheet_name="Sheet1")  
    df2.to_excel(writer, sheet_name="Sheet2")

In [2]:
# 用 ExcelWriter 设置日期格式 和 时期时间格式
from datetime import date, datetime
import pandas as pd
df = pd.DataFrame(
    [
        [date(2025, 8, 30), date(1999, 9, 24)],
        [datetime(1998, 5, 26, 23, 33, 4), datetime(2025, 8, 30, 15, 55, 30)],
    ],
    index=["Date", "Datetime"],
    columns=["X", "Y"]
)

with pd.ExcelWriter(
    "path_to_file.xlsx",
    date_format="YYYY-MM-DD",
    datetime_format="YYYY-MM-DD HH:MM:SS"
) as writer:
    df.to_excel(writer)

In [3]:
import pandas as pd
# 向现有 Excel 文件追加内容:
with pd.ExcelWriter("path_to_file.xlsx", mode="a", engine="openpyxl") as writer:
    df.to_excel(writer, sheet_name='Sheet3')

In [7]:
# if_sheet_exists 参数替换已经存在的工作表
with pd.ExcelWriter(
        "output(多个表).xlsx", 
        mode="a", 
        engine="openpyxl",
        if_sheet_exists="replace"
) as writer:
    df.to_excel(writer, sheet_name='Sheet1')

In [10]:
# 想同一张工作表写入多个DataFrame(横向并列)，注意if_sheet_exists 参数需设置为 overlay
with pd.ExcelWriter(
    "path_to_file.xlsx",
    mode="a", engine="openpyxl",
    if_sheet_exists="overlay"
) as writer:
    df1.to_excel(writer, sheet_name='Sheet1')
    df2.to_excel(writer, sheet_name='Sheet1', startcol=3)

In [12]:
# 将 Excel 文件存储在内存中
import io

df = pd.DataFrame([["ABC", "XYZ"]], columns=["Foo", "Bar"])
# io.BytesIO() 创建了一个内存缓冲区（类似一个临时的内存文件），用于在内存中存储数据，而不是写入到硬盘,在内存缓冲区，查看不了
buffer = io.BytesIO()

with pd.ExcelWriter(buffer) as writer:
    df.to_excel(writer)
    
# 将内存中的数据写入硬盘，生成物理文件，可以查看生成的"output11"
with open("output11.xlsx", "wb") as f:
    f.write(buffer.getvalue())  # buffer.getvalue() 获取内存中的二进制数据

In [13]:
# 将 Excel 文件打包到 zip 压缩文件
import zipfile

df = pd.DataFrame([["ABC", "XYZ"]], columns=["Foo", "Bar"])

with zipfile.ZipFile("path_to_file.zip", "w") as zf:  # 文件夹
    with zf.open("filename.xlsx", "w") as buffer:  # 压缩文件内的excel文件
        with pd.ExcelWriter(buffer) as writer:
            df.to_excel(writer)

In [15]:
# 向底层引擎传递额外的参数
with pd.ExcelWriter(
        "path_to_file.xlsx",
        engine="xlsxwriter",
        engine_kwargs={"options": {"nan_inf_to_errors": True}}
) as writer:
    df.to_excel(writer)

In [16]:
# 再追加模式下，engine_kwags会传递给openpyxl的load_workbook
with pd.ExcelWriter(
        "path_to_file.xlsx",
        engine="openpyxl",
        mode='a',
        engine_kwargs={"keep_vba": True}
) as writer:
    df.to_excel(writer, sheet_name='Sheet2')